src/lib.rs

```diff
@@ -184,16 +173,10 @@ where
     );
     let begin_transaction: BeginTransactionResponse = builder.execute()?;
     let is_insert = !entity.has_key();
-    let key = entity
-        .key()
-        .unwrap_or_else(|| -> Key { generate_empty_key(kind) });
-    let properties: DatastoreProperties = DatastoreProperties::from(entity)
-        .ok_or(DatastoreParseError::NoProperties)?;
-
-    let ent = Entity {
-        key: Some(key),
-        properties: Some(properties.into_map()),
-    };
+    let mut ent: Entity = entity.try_into()?;
+    if !ent.key.is_some() {
+        ent.key = Some(generate_empty_key(kind));
```

Comment:
generate_key should be able to be a `const fn` :)

Reply:
ok, not really sure about `const fn` and exactly when they can be used. Know we have talked about them, but I have never put that into practice. Will add `const fn` to my read list and change where possible

Reply:
nope!

```
error[E0723]: heap allocations are not allowed in const fn
```